### ISSUE TITLE
SSH agent errors blocking every tsh command

### ISSUE THREAD

**Issue by drewwells** [gravitational/teleport#22326]
I run `1password` ssh-agent SSH'd into a linux machine. If I turn off `ssh-agent`, I am able to use `tsh`. I don't know what keys it's trying to shove into `SSH`, but it would be great to simply disable the feature rather than being unable to use teleport at all right now.

Expected behavior:
Clusters are listed
```
-> % tsh clusters
```

Current behavior:
```
-> % tsh clusters
ERROR: failed to add one or more keys to the agent.
agent: failure, agent: failure
```

Bug details:
- Teleport version 12.0.2
- Recreation steps: run any `tsh` command
- Debug logs
```

ERROR REPORT:
Original Error: trace.aggregate agent: failure, agent: failure
Stack Trace:
        github.com/gravitational/teleport/lib/client/keyagent.go:233 github.com/gravitational/teleport/lib/client.(*LocalKeyAgent).LoadKey
        github.com/gravitational/teleport/lib/client/keyagent.go:176 github.com/gravitational/teleport/lib/client.(*LocalKeyAgent).LoadKeyForCluster
        github.com/gravitational/teleport/lib/client/api.go:1032 github.com/gravitational/teleport/lib/client.(*TeleportClient).LoadKeyForCluster
        github.com/gravitational/teleport/tool/tsh/tsh.go:3367 main.makeClientForProxy
        github.com/gravitational/teleport/tool/tsh/tsh.go:3116 main.makeClient
        github.com/gravitational/teleport/tool/tsh/config.go:45 main.onConfig
        github.com/gravitational/teleport/tool/tsh/tsh.go:1165 main.Run
        github.com/gravitational/teleport/tool/tsh/tsh.go:480 main.main
        runtime/proc.go:250 runtime.main
        runtime/asm_amd64.s:1594 runtime.goexit
User Message: failed to add one or more keys to the agent.
        agent: failure, agent: failure
```

**Comment by devicenull** [comment]
I'm seeing this with Putty and pagent as well (only after upgrading to v12.0.4)

**Comment by drewwells** [comment]
Same issue in 12.0.5, wish I could turn off the ssh agent. Happens even with `tsh kube`

**Comment by programmerq** [comment]
You can set the following environment variable to disable adding keys to the agent:

```
TELEPORT_ADD_KEYS_TO_AGENT=no
```

This env var corresponds to the `tsh -k` option:

```
  -k, --add-keys-to-agent        Controls how keys are handled. Valid values are
                                 [auto no yes only].
```

https://goteleport.com/docs/reference/cli/#tsh-environment-variables

**Comment by na4ma4** [comment]
It seems like `tsh kube login` doesn't pass on `--add-keys-to-agent` to the kube config.

It seems logical that [kubeLoginCommand](https://github.com/gravitational/teleport/blob/be2a9535c7a04f66e1ed01f7e7c3480297e95e64/tool/tsh/kube.go#L1018) would check `cf.AddKeysToAgent` and pass it along to the kubeconfig update.

**Comment by LooFifteen** [comment]
I'm using 1Password 8.10.9 on Windows 10, using its custom SSH agent - I am able to reproduce this and turning off said agent allows me to continue using tsh.

**Comment by Nezteb** [comment]
I'm also using 1Password (on macOS) with its SSH agent.

Regarding the `tsh` environment variable docs:

| **Environment Variable**     | **Description**                                                             | **Example Value**   |
|------------------------------|-----------------------------------------------------------------------------|---------------------|
| TELEPORT_ADD_KEYS_TO_AGENT   | Specifies if the user certificate should be stored on the running SSH agent | yes, no, auto, only |
| TELEPORT_USE_LOCAL_SSH_AGENT | Disable or enable local SSH agent integration                               | true, false         |

I was able to use either variable to successfully log in:
- `TELEPORT_ADD_KEYS_TO_AGENT=no tsh login --auth=local --user=myusername --proxy=myproxy.com:443`
- `TELEPORT_USE_LOCAL_SSH_AGENT=false tsh login --auth=local --user=myusername --proxy=myproxy.com:443`

I'm not sure what the value of `only` does for adding the keys; I'm guessing you can pass a path to a specific key to use? I [couldn't find any code referencing those variables](https://github.com/search?q=org%3Agravitational+%28addKeysToAgentEnvVar+OR+useLocalSSHAgentEnvVar+OR+TELEPORT_ADD_KEYS_TO_AGENT+OR+TELEPORT_USE_LOCAL_SSH_AGENT%29&type=code) to confirm though. If I try using `only`, the same generic failure occurs.

I also wonder: is there a way we can get copies of the certificates being used? 1Password's SSH agent will prompt for any SSH key it recognizes, so if we can manually add it locally we can bypass the issue altogether. 😄

**Comment by okaufmann** [comment]
@Nezteb Do you also know a way to make Teleport Connect work? It has the same problems, but I could not find out how to set these variables when running the App....

![CleanShot 2024-02-15 at 00 06 34](https://github.com/gravitational/teleport/assets/4414498/b9688b85-7243-46da-af9b-226263abe966)

**Comment by injectedfusion** [comment]
> I'm using 1Password 8.10.9 on Windows 10, using its custom SSH agent - I am able to reproduce this and turning off said agent allows me to continue using tsh.
> 
> Edit: using the environment variable to disable adding keys to the agent also worked as a workaround, which I will now be using until something happens.
> 
> Edit 2: 1Password does say that adding keys is not possible when using OpenSSH: https://developer.1password.com/docs/ssh/agent/compatibility#openssh

Same here. I had to disable adding keys to the agent `--add-keys-to-agent no` as I'm using npiperelay for the 1Password SSH Agent with WSL2 on Windows 11.

**Comment by Rubik-AU** [comment]
> @Nezteb Do you also know a way to make Teleport Connect work? It has the same problems, but I could not find out how to set these variables when running the App....
> 
> ![CleanShot 2024-02-15 at 00 06 34](https://private-user-images.githubusercontent.com/4414498/304901091-b9688b85-7243-46da-af9b-226263abe966.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQyNzQzODEsIm5iZiI6MTcxNDI3NDA4MSwicGF0aCI6Ii80NDE0NDk4LzMwNDkwMTA5MS1iOTY4OGI4NS03MjQzLTQ2ZGEtYWY5Yi0yMjYyNjNhYmU5NjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MjhUMDMxNDQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWQwOWUwNmIyY2MzZDk4YTFiZjRkYmIzMjkzNGQ3YTExYzFiNzY2ZTk3MGQxNGUyYWU2Yjc3MDY0N2Q0NDI4MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.W5a3Ynt7WK4vkuFEpnfqYlDaHfSCezkwwFwm9S6hcqE)

I'm also having the same issue. Is there any work around it? Thanks.

![Screenshot 2024-04-28 131713](https://github.com/gravitational/teleport/assets/47562901/eccd0dd0-cd6f-4d5c-b917-7ff9d2ef47b3)

**Comment by ravicious** [comment]
The workaround from https://github.com/gravitational/teleport/issues/22326#issuecomment-1492588949 should still work for Teleport Connect. On Windows, try setting `TELEPORT_ADD_KEYS_TO_AGENT=no` in your environment variables [through the System Control Panel](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.4#set-environment-variables-in-the-system-control-panel).

On other systems it's a bit more tricky. Your zsh/bash session files are not read when starting a desktop application, so setting an env var there won't help. Until we add a config file option for this, your best bet is to pass the env var when starting the app from the command line.

macOS:

```
TELEPORT_ADD_KEYS_TO_AGENT=no open -a "Teleport Connect"
```

Linux:

```
TELEPORT_ADD_KEYS_TO_AGENT=no teleport-connect
```

**Comment by colossus-digital** [comment]
> The workaround from [#22326 (comment)](https://github.com/gravitational/teleport/issues/22326#issuecomment-1492588949) should still work for Teleport Connect. On Windows, try setting `TELEPORT_ADD_KEYS_TO_AGENT` to `no` in your user environment variables [through the System Control Panel](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.4#set-environment-variables-in-the-system-control-panel).
> 
> On other systems it's a bit more tricky. On macOS, your shell initialization files are not read when starting a desktop application, so setting an env var there won't help. Until we add a config file option for this, your best bet is to pass the env var when starting the app from the command line.
> 
> macOS:
> 
> ```
> TELEPORT_ADD_KEYS_TO_AGENT=no open -a "Teleport Connect"
> ```
> 
> Linux:
> 
> ```
> TELEPORT_ADD_KEYS_TO_AGENT=no teleport-connect
> ```

The workaround doesn't work for teleport connect on windows but only for tsh. 
I can't believe that this bug has not been fixed after a long time :-(

**Comment by ethan-gallant** [comment]
This issue can be reproduced by having many principals (usernames) in the certificate. 

I think there is something amiss with the buffering or formatting.

**Comment by ethan-gallant** [comment]
For our case this looks to stem from a limitation inside of OpenSSH itself. There is a 256 principal limit for SSH keys which seems to be the issue we're hitting here.

https://github.com/openssh/openssh-portable/commit/8668706d0f52654fe64c0ca41a96113aeab8d2b8

**Comment by Sayedmasudali52** [comment]
1.it’s a good according report

**Comment by ravicious** [comment]
When we were trying to debug another issue related to SSH agents (https://github.com/gravitational/teleport/issues/3169#issuecomment-2359915904), we've added a proper config option to Teleport Connect which disables adding keys to the agent.

Could you try [updating your Teleport Connect config](https://goteleport.com/docs/connect-your-client/teleport-connect/#configuration) with `"sshAgent.addKeysToAgent": "no"` and report back if it fixes the issue? This field is supported since Teleport Connect v17.0.0, v16.4.3, and v15.4.20.

It should help us learn if the previous approach simply didn't end up changing this setting to "no" where it mattered or if setting it to "no" doesn't change anything. If it still doesn't work, could you post [debug logs](https://goteleport.com/docs/connect-your-client/teleport-connect/#submitting-an-issue)?

---

From the triage I did on #3169, I think this problem might be caused by faulty SSH cert support detection logic. It checks if the `SSH_AUTH_SOCK` env var contains `"gpg-agent"`and if not, it assumes that the agent supports SSH certs. I assume this env var is not present on Windows at all, hence tsh assumes that the agent does support SSH certs and explicitly turning off adding keys to the agent resolves the issue.

https://github.com/gravitational/teleport/blob/aa34c34dad985a33d6607d602969cbf471246fa8/lib/client/keyagent.go#L89-L95

**Comment by ethan-gallant** [comment]
@ravicious should this be split into another issue in that instance?

I've determined that this particular issue on MacOS & Linux systems can be caused by having too many roles / principals assigned to a user. The SSH agent rejects the certificate with an unknown error due to the size being too large.

**Comment by ravicious** [comment]
@ethan-gallant Teleport Connect ultimately delegates all operations to tsh underneath. The workaround for the issue seems to be making sure that tsh doesn't attempt to add keys to the agent. I'd like to make sure that Windows users are able to use this workaround.

As to the root cause of the problem, it seems that my comment about `SSH_AUTH_SOCK` was preposterous, given your comment about the principal limit and https://github.com/gravitational/teleport/issues/22326#issuecomment-1686349116 about 1Password not being compatible with ssh-add when using OpenSSH.

Do you also happen to use 1Password's agent? If not, then I think it might be a good idea to create a separate issue for the principal limit and I could rename this issue to make it more specific to 1Password.

**Comment by Atroskelis** [comment]
Hello!
I assume this is a client specific issue, as running on a different host runs okay.
If so does anyone know the path to the keys saved in the Windows 11 Environment? Deleting .tsh has no effect.

**Comment by martin-westra** [comment]
I might have found a solution for Windows.
I looked at the permissions for the file `known_hosts` in `%USERPROFILE%\.ssh` and I saw there was no "Full Control" selected for the current user. I was able to edit the permissions and added "Full Control" to `known_hosts`.
Now I can set the config: `"sshAgent.addKeysToAgent": "yes"`
